### DISCUSSION TITLE
Enhance the quality of scanned PDF

### DISCUSSION THREAD

**Discussion by maiiabocharova** [pymupdf/PyMuPDF#1517] [Q&A] [answered]
Sorry to disturb, but is there a way in enhance a scanned PDF quality (make dirty greyish background white and make the contrast higher)?

**Comment by JorjMcKie** [comment, marked as answer]
You do not disturb at all, no worries.

> make dirty greyish background white and make the contrast higher

In general no, this is out of PyMuPDF's scope. It actually is an image enhancement topic. You might want to consult Pillow documentation for this.
Also, OCRmyPDF has a feature in that area if I remember correctly. Including improvement of the scanned image orientation.

There are a handful of things in PyMuPDF which may also help, e.g.
1. gamma correction of the pixmap
2. you **_could_** write your own routine and examine the `pix.samples`. Like making it white based on some threshold, e.g. if for an RGB image the pixel is greater than `(200, 200, 200)` replace it by `(255, 255, 255)` or whatever. But it is a shaky thing to do, Pillow and the tools underneath OCRmyPDFs hood are proven and probably less error prone.

**Reply by maiiabocharova** [reply]
Thanks a lot! Didn't think before about changing grey pixels to white based on some threshold using Pillow, it should work, thanks once again!

**Reply by maiiabocharova** [reply]
One more question: when I have a PIL Image which I want to make a pdf page - do I need to first save in as png and afterwards load as described here in [documentation](https://pymupdf.readthedocs.io/en/latest/faq.html#how-to-make-one-pdf-of-all-your-pictures-or-files)  or is there a way to directly use it (or a numpy array from which I got in `im = Image.fromarray(np_image)`) without first saving?

**Reply by JorjMcKie** [reply]
You do not need to save anything to disk to achieve this.
You can "save" the PIL image to memory in any desired format (or possible format, depending on what image type you have)
Then you can do one of two things:
1. Open the image in memory as a 1-page fitz.Document and convert this to a 1-page PDF, which you can append (via `.insert_pdf()`) to your final PDF.
2. In your final PDF make a new page (with appropriate width / height - probably taken from the image), and insert the image from memory.

How to convert a PIL image to something that PyMuPDF can handle:
```python
# img is the PIL Image
import io  # to use as an intermediate storage in memory
fp = io.BytesIO()  # Pillow will write to this "file pointer"
img.save(fp, "JPEG", ... Pillow more parameters ...)
# choose the desired format like JPEG, PNG, ... depending on the image you are dealing with
# other parameters might include "optimize" or whatever
```
```python
# Method 1 from above:
imgdoc = fitz.open(fp.getvalue())  # open image as a fitz.Document
imgpdf = fitz.open("pdf", imgdoc.convert_to_pdf())  # convert to PDF Documen
doc.insert_pdf(imgpdf)  # append to target Document
# etc.
# ----------------------------------
# Method 2 from above:
page = doc.new_page(width=img.size[0], height=img.size[1])  # make page in target Document
page.insert_image(page.rect, stream=fp.getvalue())
```

**Reply by maiiabocharova** [reply]
Thank you!!! For code and explanations!

**Comment by JorjMcKie** [comment]
Let me transfer this to a **_Discussion._**
